index.js

```diff
@@ -155,6 +156,65 @@ class WebTorrent extends EventEmitter {
     }
   }
 
+  /**
+   * Accepts a path to a service worker file to register, or an existing service worker registration
+   * [navigator.serviceWorker.controller] which must be activated, creates a file server for streamed
+   * file rendering to use.
+   *
+   * @param  {string|ServiceWorker} registration
+   * @param {function=} cb
+   * @return {null}
+   */
+  async registerWorker (registration, cb = () => {}) {
+    if (!('serviceWorker' in navigator)) throw new Error('Service workers not supported')
+
+    if (typeof registration === 'string') {
+      if (navigator.serviceWorker.controller !== null) throw new Error('A worker is already registered')
+      await navigator.serviceWorker.register(registration, {
+        scope: window.location.pathname.substr(0, window.location.pathname.lastIndexOf('/') + 1)
+      }).catch(e => {
+        // when loaded by redirects, workers sometimes enter an invalid state, this always happens when loaded by a protocol handler
+        // if (String(e) === 'InvalidStateError: Failed to register a ServiceWorker: The document is in an invalid state.') {
+        //   window.location.reload() // weird workaround for a weird bug
+        // }
+        throw e
+      })
+    } else {
+      if (!(registration instanceof ServiceWorker)) throw new Error('Invalid worker registration')
+      if (registration.state !== 'activated') throw new Error('Worker isn\'t activated')
+    }
+
+    this.serviceWorker = registration instanceof ServiceWorker ? registration : navigator.serviceWorker.controller
+
+    navigator.serviceWorker.addEventListener('message', event => {
+      if (!event.data.url) return null
+      let [infoHash, ...filePath] = event.data.url.slice(event.data.url.indexOf(event.data.scope + 'webtorrent/') + 11 + event.data.scope.length).split('/')
+      filePath = decodeURI(filePath.join('/'))
+      if (!infoHash || !filePath) return null
+
+      const [port] = event.ports
+
+      const file = this.get(infoHash) && this.get(infoHash).files.find(file => file.path === filePath)
+      if (!file) return null
```

Comment:
I think it would be nice to provide the page with some information about the request...    like a 404 response.

Reply:
general issue here, is that you can have multiple webtorrent instances running of off a single service worker, and only one wt instance can be the controller, so some instances might have the files, some might not, that's why I didn't do any error handling for this

Reply:
Ok, I see that...  but that's a pretty complex use-case.    It's going to be up to the WT maintainers what they want, but if it were me, I'd vote for the simple impl.   The page requesting the url is the page that owns the wt instance.

logic in fetch handler could be cleaned up a little to message the requesting client.  Not loop over all of them.
I don't know how common a popout video player window would be.
I don't personally have that use case.   But that's where others could implement the interfaces we should define.

How much do we want to try to support in the default impl?  you'll notice that wt doesn't have many Store impls by default.